lib/charms/prometheus_k8s/v0/prometheus_remote_write.py

```diff
@@ -681,9 +684,37 @@ def _push_alerts_to_relation_databag(self, relation: Relation) -> None:
 
         alert_rules_as_dict = alert_rules.as_dict()
 
+        alert_rules_as_dict = self._parse_and_add_extra_alerts(alert_rules_as_dict)
+
         if alert_rules_as_dict:
             relation.data[self._charm.app]["alert_rules"] = json.dumps(alert_rules_as_dict)
 
+    def _parse_and_add_extra_alerts(self, alert_rules_as_dict: dict) -> dict:
+        """Parse alerts from `self.extra_alerts_callable`.
+
+        Call `self.extra_alerts_callable`, parse alert groups and
+        extend (mutate) `alert_rules_as_dict` with new alerts.
+
+        Returns:
+            (dict) Modified `alert_rules_as_dict`. Return is done only for explicit mutation.
+        """
+        if callable(self.extra_alerts_callable):
+            extra_alerts_list = []
+            extra_alerts = self.extra_alerts_callable()
+            if extra_alerts:
+                for topology_identifier, alert_rule_groups in extra_alerts.items():
+                    extra_alerts_list.extend(alert_rule_groups.get("groups", []))
```

Comment:
This still seems kind of unnecessary.

We could do:

```python
for alert_rule_groups in extra_alerts.values():
  extra_alerts_list.extend(...)
```

It could also be a comprehension, but even I'll admit that would be hard to reason about if it had to be extended/changed later.

Reply:
valid point, thanks!